### PR TITLE
kvserver: split snapshot SSTables for mvcc keys into multiple SSTs

### LINKED ISSUE #67284: kv/kvserver: split incoming snapshot user keys into multiple sstables
Currently, when receiving a snapshot, we ingest a fixed number of sstables, corresponding to the range's various contiguous keyspaces. When the default range size increased from 64 MB to 512 MB, we started ingesting user data sstables up to 512 MB. These large files cause more expensive compactions.

Here's a [relevant TODO](https://github.com/cockroachdb/cockroach/blob/9ba3198cbdfd856c429bde695ffc1f0ec0f96c37/pkg/kv/kvserver/store_snapshot.go#L241).

Related thread: https://cockroachlabs.slack.com/archives/CAC6K3SLU/p1686237663708959?thread_ts=1686084258.076069&cid=CAC6K3SLU

Related to cockroachdb/pebble#1181.

cc @sumeerbhola

Jira issue: CRDB-8471

### PR DESCRIPTION
**Pull request by itsbilal**
Previously, we'd only create one sstable for all mvcc keys in a range when ingesting a rebalance/recovery snapshot into Pebble. This increased write-amp in Pebble as more sstables would have to be compacted into it (or the sstable then split into smaller ones in Pebble), and had other consequences such as massive filter blocks in the large singular sstable.

This change adds a new cluster setting,
kv.snapshot_rebalance.max_sst_size, that sets the max size of the sstables containing user/mvcc keys in a range. If an sstable exceeds this size in multiSSTWriter, we roll over that sstable and create a new one.

Epic: CRDB-8471
Fixes: #67284

Release note (performance improvement): Reduce the write-amplification impact of rebalances by splitting snapshot sstable files into smaller ones before ingesting them into Pebble.

### CODE REVIEW

**Comment by itsbilal** [comment]
This PR is actually missing the range key fragmenting piece. I'll change it to a draft PR as we'll need that.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @jbowens)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZBv1AB2PtPxYAXX0:-O39ZBv26VIktGDobB-L:br2txr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L270)):*
> ```Go
> 		// Finish the current SST, write to the file, and move to the next key
> 		// range.
> 		if err := msstw.finalizeSST(ctx); err != nil {
> ```

[nit] It would be nice if we would pass `key` to `finalizeSST` so it can assert that the sst's largest key is strictly smaller than `key`
___
*[`pkg/kv/kvserver/store_snapshot.go` line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39_4GEDSj4Mq09f8Xh:-O39_4GEDSj4Mq09f8Xi:b-ujwtfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L285)):*
> ```Go
> 		return errors.AssertionFailedf("client error: expected %s to fall in one of %s or %s", key, msstw.localKeySpans, msstw.mvccKeySpans)
> 	}
> 	if msstw.currSpanIsMVCCSpan() && msstw.maxSSTSize > 0 && msstw.currSST.DataSize > msstw.maxSSTSize {
> ```

How does this work if we already did a PutInternalRangeDelete or PutInternalRangeKey that ends after the current key? Feels like we need to keep track of that watermark and never split before it.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZlDf-TqlpI0Md5U5:-O39ZlDf-TqlpI0Md5U6:bl39e46) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L288)):*
> ```Go
> 		// We're in an MVCC / user keys span, and the current sstable has exceeded
> 		// the max size for MVCC sstables that we should be creating. Split this
> 		// sstable into smaller ones. We do this by splitting the mvccKeySpan into
> ```

[nit] remove "into"
___
*[`pkg/kv/kvserver/store_snapshot.go` line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39Xw4x6m_7TxhLA4bW:-O39Xw4x6m_7TxhLA4bX:b-vi1k04) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L299)):*
> ```Go
> 		newSpan := roachpb.Span{Key: currSpan.EndKey, EndKey: oldEndKey}
> 		msstw.mvccKeySpans = append(msstw.mvccKeySpans, newSpan)
> 		if msstw.currSpan < len(msstw.localKeySpans)+len(msstw.mvccKeySpans)-1 {
> ```

[nit] you can use `slices.Insert(mvccKeySpans, curSpan, newSpan)`

**Review by jbowens — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EMc-q2iFzQSJ5Q2B4:-O3EMc-q2iFzQSJ5Q2B5:b-jbp0hz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L71)):*
> ```Go
> 	"kv.snapshot_rebalance.max_sst_size",
> 	"maximum size of a rebalance or recovery SST size",
> 	32<<20, // 32 MB
> ```

nit: should we make this 128 MB to match our expectation that these tables will be ingested into L6?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EQ8vLAodCYHG7Hmw0:-O3EQ8vLAodCYHG7Hmw1:b-79ykp3) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L150)):*
> ```Go
> 	// having this bool to true also disables all range key fragmentation
> 	// and splitting of sstables in the mvcc span.
> 	skipRangeDelForMVCCSpan bool
> ```

nit: call it `skipClearForMVCCSpan`or the like so that it's not suggesting that we're writing a range deletion but not a range key deletion
___
*[`pkg/kv/kvserver/store_snapshot.go` line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EPMhvEsOxVBfAW0La:-O3EPMhvEsOxVBfAW0Lb:b81yfig) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L176)):*
> ```Go
> 		skipRangeDelForMVCCSpan: skipRangeDelForMVCCSpan,
> 	}
> 	if st != nil && !skipRangeDelForMVCCSpan {
> ```

do any callers actually pass in a nil `st` and is it possible to update the caller instead?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EOiVz0TXbigOLPRS1:-O3EOiVz0TXbigOLPRS2:b4zce2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L179)):*
> ```Go
> 		// If skipRangeDelForMVCCSpan is true, we don't split the MVCC span across
> 		// multiple sstables, as addRangeDelForMVCCSpan could be called by the caller
> 		// at any time.
> ```

I didn't understand the meaning of this "addRangeDelForMVCCSpan could be called by the caller at any time."

Since we write range deletions even when performing an excise now, when wouldn't we add a range deletion?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EPnViCsEdJCINSLBn:-O3EPnViCsEdJCINSLBo:by468h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L231)):*
> ```Go
> 		// we're writing.
> 		if !msstw.skipRangeDelForMVCCSpan {
> 			if err := msstw.currSST.ClearRawRange(
> ```

how does this interact with range key fragmentation since `ClearRawRange` also writes a `RangeKeyDelete`

**Comment by itsbilal** [comment]
TFTRs! I'll be incorporating the feedback soon. Might need to rework a decent chunk of my approach here as we can't fragment at any EngineKeys due to expectations from Pebble around range keys having suffixless keys at the external (ingestion) point, even though fragmentation within Pebble of range keys at suffixed keys is allowed.

That plus other instances of nuances means a lot of this PR is going to remain in flux for the next couple days, so I'd encourage holding off on more reviews until I've marked it as ready-for-review. Thanks!

**Review by itsbilal — COMMENTED** [review]
TFTRs! This is ready for a look now. I've filed https://github.com/cockroachdb/pebble/issues/3829 based on a conversation with Jackson about how this change veers very close to some missing assertions around range key invariants in Pebble on the ingestion path.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-:-O3dP9Dz4F-zwCDKcrDP:b-m0pyt9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZBv1AB2PtPxYAXX0:-O3dOmHd3o1ma-f-FbuW:b-3v0j5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L270)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] It would be nice if we would pass `key` to `finalizeSST` so it can assert that the sst's largest key is strictly smaller than `key`
</blockquote></details>

We get this sort of checking for free in the ingestSortAndVerify check on the ingestion path, but also I'm not aware if the sst writer keeps track of the largest key added to it. Should we add it still?
___
*[`pkg/kv/kvserver/store_snapshot.go` line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39_4GEDSj4Mq09f8Xh:-O3dOTFxBhqwPwEBN8_M:b-7vdfct) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L285)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

How does this work if we already did a PutInternalRangeDelete or PutInternalRangeKey that ends after the current key? Feels like we need to keep track of that watermark and never split before it.
</blockquote></details>

Both of those methods have a panic if we call them with `!skipClearForMVCCSpan`. And in the `skipClearForMVCCSpan` case, we don't do sstable splits, so we're covered.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 288 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZlDf-TqlpI0Md5U5:-O3dOS_uEgUgslyjRot7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L288)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] remove "into"
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 299 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39Xw4x6m_7TxhLA4bW:-O3dOQxCEDfvXxlijNvG:b-xc66z8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L299)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] you can use `slices.Insert(mvccKeySpans, curSpan, newSpan)`
</blockquote></details>

Done (unnecessary as this code got reworked)
___
*[`pkg/kv/kvserver/store_snapshot.go` line 71 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EMc-q2iFzQSJ5Q2B4:-O3dNJfH64fkT5RpP4S8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L71)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: should we make this 128 MB to match our expectation that these tables will be ingested into L6?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 150 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EQ8vLAodCYHG7Hmw0:-O3dNKgV6K4WQ1AtX978:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L150)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: call it `skipClearForMVCCSpan`or the like so that it's not suggesting that we're writing a range deletion but not a range key deletion
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 176 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EPMhvEsOxVBfAW0La:-O3dO7kw8jSnyWf6x8ij:b6o1i4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L176)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

do any callers actually pass in a nil `st` and is it possible to update the caller instead?
</blockquote></details>

Done. I don't think any callers were passing it in.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 179 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EOiVz0TXbigOLPRS1:-O3dO9OOA3Z5zwUz4dak:b-b6rx7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L179)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I didn't understand the meaning of this "addRangeDelForMVCCSpan could be called by the caller at any time."

Since we write range deletions even when performing an excise now, when wouldn't we add a range deletion?
</blockquote></details>

Disaggregated storage is the only case where we wouldn't add range deletions.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O3EPnViCsEdJCINSLBn:-O3dOCsd1nkmZseEk7PZ:b-p0tsv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa587b68c557d07d7d74ef8e2d95c8438cb2e2a8/pkg/kv/kvserver/store_snapshot.go#L231)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

how does this interact with range key fragmentation since `ClearRawRange` also writes a `RangeKeyDelete`
</blockquote></details>

Done. Reworked this code so it only adds a point key rangedel, and we explicitly fragment the rangekeydel in the same fragmenter as all other range keys.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @itsbilal and @jbowens)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZBv1AB2PtPxYAXX0:-O3g7mpH33YgSiwIKfmf:b-493q8i) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L270)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

We get this sort of checking for free in the ingestSortAndVerify check on the ingestion path, but also I'm not aware if the sst writer keeps track of the largest key added to it. Should we add it still?
</blockquote></details>

The largest key is readily available in the sst writer Metadata() once we finish it.

**Review by itsbilal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jbowens and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 270 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O39ZBv1AB2PtPxYAXX0:-O3i0TPK3B7S0WS0zIXo:b-s4589) ([raw file](https://github.com/cockroachdb/cockroach/blob/2abdf2485ba277a69ba8152c253fa7963ee8fa69/pkg/kv/kvserver/store_snapshot.go#L270)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

The largest key is readily available in the sst writer Metadata() once we finish it.
</blockquote></details>

Ah right. Fixed!

**Review by RaduBerinde — APPROVED** [review]
:lgtm: but I'd wait for an OK from Jackson too as I'm not familiar with much of this code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-:-O4BnWF1EZqAAs7brB9A:by78isc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @itsbilal and @jbowens)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bkcr77xSu89aNxeW4:-O4Bkcr77xSu89aNxeW5:b5v4w9g) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L209)):*
> ```Go
> 	for i := range key.Keys {
> 		if err := msstw.currSST.PutInternalRangeKey(key.Start, key.End, key.Keys[i]); err != nil {
> 			panic("failed to put range key in sst")
> ```

[nit] print the error too
___
*[`pkg/kv/kvserver/store_snapshot.go` line 219 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4BktT6E8OH1mZan4Ih:-O4BktT6E8OH1mZan4Ii:b36uqs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L219)):*
> ```Go
> func (msstw *multiSSTWriter) currentSpan() roachpb.Span {
> 	if msstw.currSpan >= len(msstw.localKeySpans) {
> 		if msstw.currSpan >= len(msstw.localKeySpans)+len(msstw.mvccSSTSpans) {
> ```

[nit] maybe move this check inside `currSpanIsMVCCSpan` and use that method here
___
*[`pkg/kv/kvserver/store_snapshot.go` line 238 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bl-h29pOzMw35hRmo:-O4Bl-h29pOzMw35hRmp:b6ggka7) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L238)):*
> ```Go
> 	newSST := storage.MakeIngestionSSTWriter(ctx, msstw.st, newSSTFile)
> 	msstw.currSST = newSST
> 	if msstw.currSpan < len(msstw.localKeySpans) || (!msstw.skipClearForMVCCSpan && msstw.currSpan <= len(msstw.localKeySpans)) {
> ```

`if !curSpanIsMVCCSpan() ||`
___
*[`pkg/kv/kvserver/store_snapshot.go` line 240 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4BlElh-sayzs_i-2P2:-O4BlElh-sayzs_i-2P3:b-ultgcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L240)):*
> ```Go
> 	if msstw.currSpan < len(msstw.localKeySpans) || (!msstw.skipClearForMVCCSpan && msstw.currSpan <= len(msstw.localKeySpans)) {
> 		// We're either in a local key span, or we're in the first MVCC sstable
> 		// span (before any splits). Add a RangeKeyDel for the whole MVCC span
> ```

"for the whole span. If this is the first MVCC span, we don't need to keep re-adding it.."
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bn-ftA7VqpRY3h011:-O4Bn-fuAgA-APUWTNHb:b-lli2xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L362)):*
> ```Go
> 	_, err = multiSSTWriter.Finish(ctx)
> 	require.NoError(t, err)
> 	require.Greater(t, len(scratch.SSTs()), len(keySpans))
> ```

[nit] Ideally we would check the contents of the ssts too - it could be a datadriven test where we input the keys written and then we output the contents of the resulting SSTs.

We should also test the use of `ClearEngineRange` and `skipRangeDelForMVCCSpan=true`

**Review by itsbilal — COMMENTED** [review]
TFTR! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-:-O4Cc-2M2Xeq_KB7cnEh:bz5m8k4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jbowens and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bkcr77xSu89aNxeW4:-O4CPhkNBkdW_9b1-7jZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L209)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] print the error too
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 238 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bl-h29pOzMw35hRmo:-O4CQODNBMofQ7ZVtCmm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L238)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`if !curSpanIsMVCCSpan() ||`
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/store_snapshot.go` line 240 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4BlElh-sayzs_i-2P2:-O4CQOj34UhPeTR64xkv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/store_snapshot.go#L240)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

"for the whole span. If this is the first MVCC span, we don't need to keep re-adding it.."
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_sst_snapshot_storage_test.go` line 362 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4Bn-ftA7VqpRY3h011:-O4CXYSj8EcU3XHv2JW8:blcgz01) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb1a9d573edcf24802046bed5948d2ed868aa2af/pkg/kv/kvserver/replica_sst_snapshot_storage_test.go#L362)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Ideally we would check the contents of the ssts too - it could be a datadriven test where we input the keys written and then we output the contents of the resulting SSTs.

We should also test the use of `ClearEngineRange` and `skipRangeDelForMVCCSpan=true`
</blockquote></details>

The `skipRangeDelForMVCCSpan` case disables splitting and is already tested in `TestMultiSSTWriterAddLastSpan`, at least. I updated this test to also verify the contents of the SSTs though!

**Review by jbowens — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-:-O4Dk0U42T5YhQMDJDmP:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @itsbilal and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 134 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4DBfUP-BKV52DOnr9p:-O4DBfUP-BKV52DOnr9q:be13jxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c785b252277f728afe6d12bc601ffd592c96564/pkg/kv/kvserver/store_snapshot.go#L134)):*
> ```Go
> 	currSST storage.SSTWriter
> 	// localKeySpans are key spans that are considered unsplittable across sstables.
> 	// mvccKeySpan can be split across multiple sstables if one of them exceeds
> ```

super nit: maybe precede this with explaining that localKeySpans define the range's range local key spans.  this comment makes it seem like the defining characteristic is that we won't split them.

also, may precede the second sentence with "In contrast mvccKeySpan can..."

**Review by itsbilal — COMMENTED** [review]
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-:-O4GlQ7f67OpIQhNu93k:bz5mair)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @jbowens and @RaduBerinde)

___
*[`pkg/kv/kvserver/store_snapshot.go` line 134 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/127997#-O4DBfUP-BKV52DOnr9p:-O4GlPoF9G3_RdY3O7sU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c785b252277f728afe6d12bc601ffd592c96564/pkg/kv/kvserver/store_snapshot.go#L134)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

super nit: maybe precede this with explaining that localKeySpans define the range's range local key spans.  this comment makes it seem like the defining characteristic is that we won't split them.

also, may precede the second sentence with "In contrast mvccKeySpan can..."
</blockquote></details>

Done.

**Comment by itsbilal** [comment]
bors r=jbowens

**Pull request merged**
